OrbitGl/TimerInfosIterator.h

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBITGL_TIMER_INFOS_ITERATOR_H_
+#define ORBITGL_TIMER_INFOS_ITERATOR_H_
+
+#include <memory>
+#include <vector>
+
+#include "TimerChain.h"
```

Comment:
I am thinking that having _TimerInfosIterator_ dependant on _TimerChain_ (which depends on _TextBox_) will make not possible to move this class outside OrbitGl within the serialisation, so the _Write_ method could not be moved to be independent from OrbitGl. Or am I missing something? Are _TimerChain_ and _TextBox_ meant to be moved outside _OrbitGl_ as well? Would that be possible?

Reply:
Maybe _TimerInfosIterator_ should contain _TimerInfo_ instead of _TimerChain_ as checking the old code of where timers are being written, I see that _TimerInfo_ it is the type of data that ultimately is being written: the _TimerChains_ are retrieved, then from these, a _TimerBlock_ , that is used to get and write the _TimerInfo_

Reply:
So if we get those _TimerInfo_ beforehand and pass it to the write method, the serialiser would receive all the data it needs and can be completely independent from _TimerChain_ (and therefore _TextBox_ and _OrbitGl_)

Reply:
For clarity, the piece of code I mean is:

```
// Timers
int writes_count = 0;
  std::vector<std::shared_ptr<TimerChain>> chains = time_graph_->GetAllTimerChains();
  for (auto& chain : chains) {
    if (!chain) continue;
    for (TimerChainIterator it = chain->begin(); it != chain->end(); ++it) {
      TimerBlock& block = *it;
      for (uint32_t k = 0; k < block.size(); ++k) {
        WriteMessage(&block[k].GetTimerInfo(), &coded_output);
        if (++writes_count > timers_count) {
          return;
        }
      }
    }
  }
```
So the important bit I refer to is `WriteMessage(&block[k].GetTimerInfo(), &coded_output);`,

Reply:
The idea would be that this TimerInfosIterator is highly dependent from OrbitGl. On the Serializer it is then only a generic template. You can call "Write(..., vector.begin(), vector.end())" in your case and would be fine.

Reply:
So that Iterator basically should just define how to iterate of the current state inside orbitgl's timer storage. But serializer remains independent.

Reply:
Ahhh, ok! Got it, makes sense, thanks!

Reply:
Welcome.